File: src/main/java/io/pivotal/project/ProjectService.java
package io.pivotal.project;

import io.pivotal.project.burndown.BurndownCsvParser;
import io.pivotal.project.burndown.BurndownPeriod;
import io.pivotal.project.burndown.BurndownProjectionModel;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.toList;

@Service
class ProjectService {

    private ProjectRepository projectRepository;
    private BurndownCsvParser burndownCsvParser;

    @Autowired
    public ProjectService(ProjectRepository projectRepository, BurndownCsvParser burndownCsvParser) {
        this.projectRepository = projectRepository;
        this.burndownCsvParser = burndownCsvParser;
    }

    List<Project> getAllProjects() {
        return projectRepository
            .getAllProjectEntities()
            .stream()
            .map(Project::new)
            .collect(toList());
    }

    Optional<Project> getProjectById(int projectId) {
        ProjectEntity projectEntityById = projectRepository.getProjectEntityById(projectId);

        if (projectEntityById == null) {
            return Optional.empty();
        }

        List<BurndownPeriod> burndown = getBurndown(projectEntityById);

        Integer numberOfEmployees =
                (burndown.size() > 0) ?
                burndown.get(burndown.size()-1).getNumberOfEmployees():
                Project.DEFAULT_NUMBER_OF_EMPLOYEES;


        Project project = Project.builder()
            .id(projectEntityById.getId())
            .name(projectEntityById.getName())
            .hourlyRate(projectEntityById.getHourlyRate())
            .startDate(projectEntityById.getStartDate())
            .budget(projectEntityById.getBudget())
            .burndown(burndown)
            .numberOfEmployees(numberOfEmployees)
            .build();

        return Optional.of(project);
    }

    private List<BurndownPeriod> getBurndown(ProjectEntity projectEntityById) {
        Optional<List<BurndownPeriod>> maybeBurndown = Optional.ofNullable(burndownCsvParser.getBurndownForProjectEntity(projectEntityById));
        return maybeBurndown.orElse(Collections.emptyList());
    }

    Project saveProject(Project project) {
        ProjectEntity projectEntity = ProjectEntity.builder()
            .name(project.getName())
            .startDate(project.getStartDate())
            .hourlyRate(project.getHourlyRate())
            .budget(project.getBudget())
            .build();

        ProjectEntity savedProjectEntity = projectRepository.save(projectEntity);

        return Project.builder()
            .id(savedProjectEntity.getId())
            .name(savedProjectEntity.getName())
            .hourlyRate(savedProjectEntity.getHourlyRate())
            .startDate(savedProjectEntity.getStartDate())
            .budget(savedProjectEntity.getBudget())
            .burndown(Collections.emptyList())
            .build();
    }

    public void buildProjection(Project project, Integer numberOfEmployees) {
        if(numberOfEmployees > 0){
            project.setNumberOfEmployees(numberOfEmployees);
        }
        BurndownProjectionModel.buildOut(project);
    }
}

File: src/main/java/io/pivotal/project/ProjectService.kt
package io.pivotal.project

import io.pivotal.project.burndown.BurndownParser
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import java.util.Optional

@Service
internal open class ProjectService @Autowired
constructor(private val projectRepository: ProjectRepository, private val burndownParser: BurndownParser) {

   fun allProjects() : List<Project> =
                projectRepository
                .allProjectEntities()
                        .toList()
                .map{ Project(it) }

    fun getProjectById(projectId: Int): Project? {
        val projectEntityById = projectRepository.getProjectEntityById(projectId) ?: return null

        val burndown = getBurndown(projectEntityById)

        val project = Project(
                id = projectEntityById.id,
                name = projectEntityById.name,
                hourlyRate = projectEntityById.hourlyRate,
                startDate = projectEntityById.startDate,
                budget = projectEntityById.budget,
                burndown = burndown
                )

        return project
    }

    private fun getBurndown(projectEntityById: ProjectEntity): List<Float> {
        val maybeBurndown = Optional.ofNullable(burndownParser.getBurndownForProjectEntity(projectEntityById))
        return maybeBurndown.orElse(emptyList<Float>())
    }

    fun saveProject(project: Project): Project {
        val projectEntity = ProjectEntity(
                name = project.name,
                startDate = project.startDate,
                hourlyRate = project.hourlyRate,
                budget = project.budget)


        val savedProjectEntity = projectRepository.save(projectEntity)

        return Project(
                id = savedProjectEntity.id,
                name = savedProjectEntity.name,
                hourlyRate = savedProjectEntity.hourlyRate,
                startDate = savedProjectEntity.startDate,
                budget = savedProjectEntity.budget,
                burndown = mutableListOf<Float>())
    }
}
